### PR TITLE
add reach def in dfa

### PR DESCRIPTION
**Pull request by al3xliu**

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                (RegularTransferResult<ReachDefinitionValue, ReachDefinitionStore>)
+                        super.visitAssignment(n, p);
+        processDefinition(n, transferResult.getRegularStore());
+        return transferResult;
+    }
+
+    /**
+     * Update the information of reach definition from an assignment statement.
+     *
+     * @param def the def that should be put into the store
```
Comment (zcai1) [inline dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java:48]:
```suggestion
     * @param def the definition that should be put into the store
```
I think we should avoid using abbreviation in the explanation.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     * @param store the reach defination store
+     */
+    private void processDefinition(AssignmentNode def, ReachDefinitionStore store) {
+        store.killDef(new ReachDefinitionValue(def));
```
Comment (zcai1) [inline dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java:52]: Maybe you don't need to pass a new value. I think `def.getTarget` is sufficient for removing the previous defs.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Remove the information of a reach definition from the reach definition set.
+     *
+     * @param def a reach definition
+     */
+    public void killDef(ReachDefinitionValue def) {
+        Iterator<ReachDefinitionValue> it = reachDefSet.iterator();
+        while (it.hasNext()) {
+            ReachDefinitionValue existedDef = it.next();
+            if (existedDef.defs.getTarget().toString().equals(def.defs.getTarget().toString())) {
```
Comment (zcai1) [inline dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java:44]: By passing a node to the assignment target to this method, you can simply compare `existedDef.defs.getTarget().equals(target)` here.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return new ReachDefinitionStore(new HashSet<>(reachDefSet));
+    }
+
+    @Override
+    public ReachDefinitionStore leastUpperBound(ReachDefinitionStore other) {
+        Set<ReachDefinitionValue> reachDefSetLub =
+                new HashSet<>(this.reachDefSet.size() + other.reachDefSet.size());
+        reachDefSetLub.addAll(this.reachDefSet);
+        reachDefSetLub.addAll(other.reachDefSet);
+        System.out.println(reachDefSetLub.toString());
```
Comment (zcai1) [inline dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java:84]: Do we still need the println here?

File: dataflow/tests/reach-definition/Test.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// This file may not be renamed; it has to have the same filename as ../issue3447/Test.java .
```
Comment (zcai1) [inline dataflow/tests/reach-definition/Test.java:1]: Maybe add some extra test files? I think we also need an example under `dataflow/manual/examples`.

File: docs/manual/contributors.tex [hunk 6]
```diff
@@ -123,4 +124,5 @@
 Vlastimil Dort,
 Weitian Xing,
 Werner Dietl,
-Zhiping Cai.
+Zhiping Cai,
+alex.
```
Comment (zcai1) [inline docs/manual/contributors.tex:128]: I think you've added "Alex Liu," above.

**Review by zcai1 — COMMENTED** [review]
Thank you for the PR! I think most of the PR looks good and there are also a few comments left.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java (hunk 2 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefTransfer.java:52]: Got it.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java (hunk 3 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java:44]: Got it.

File: dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java (hunk 4 above)
Comment (al3xliu) [reply dataflow/src/main/java/org/checkerframework/dataflow/reachdefinition/ReachDefinitionStore.java:84]: Got it.

File: dataflow/tests/reach-definition/Test.java (hunk 5 above)
Comment (al3xliu) [reply dataflow/tests/reach-definition/Test.java:1]: Ok, I think I will add one more test file in this pr and a new test file in my next pr when I remove StringConcatenateAssignmentNode.

File: dataflow/build.gradle [hunk 7]
```diff
@@ -94,6 +94,47 @@ task liveVariableTest(dependsOn: [assemble, compileTestJava], group: 'Verificati
     }
 }
 
+
+task reachDefinitionTest(dependsOn: [assemble, compileTestJava], group: 'Verification') {
```
[the remark this answers is not in the document]
Comment (al3xliu) [reply dataflow/build.gradle:98]: Thanks for reviewing my code!

**Comment by al3xliu** [comment]
Move to #181 , close this PR.

**Pull request closed without merging**